### PR TITLE
add 2pc

### PR DESCRIPTION
**Pull request by morefreeze**
@disksing PTAL

### CODE REVIEW

File: scan.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+				ret[idx] = &resultRow{
+					key:         item.GetKey(),
+					value:       nil,
+					isLocked:    true,
+					primary:     item.GetPrimaryLock(),
+					lockVersion: item.GetLockVersion(),
+				}
+			default:
+				// TODO(freeze): how to handle retry.
+				log.Warn("This key will be missing because error: %s", item.GetErr())
```
Comment (disksing) [inline scan.go]: This case will not happen if server side is correctly implemented. I think we may return error here.

File: snapshot.go [hunk 2]
```diff
@@ -16,18 +16,16 @@ package ticlient
...
 
 var (
 	_ kv.Snapshot = (*tikvSnapshot)(nil)
-	_ kv.Iterator = (*tikvIter)(nil)
 )
 
 const (
-	tikvBatchSize        = 1000
+	tikvBatchSize        = 100
 	tikvScanLimit uint32 = 1000
```
Comment (disksing) [inline snapshot.go]: Unused variable?

**Comment by ngaut** [comment]
PTAL @c4pt0r @BusyJay

File: snapshot.go [hunk 3]
```diff
@@ -77,110 +76,103 @@ func (s *tikvSnapshot) Get(k kv.Key) ([]byte, error) {
...
+//if !it.valid {
+//return nil
+//}
+//return it.rs[it.idx].value
+//}
+
+//// Close close iterator.
+//func (it *tikvIter) Close() {
+//it.valid = false
+//}
```
Comment (disksing) [inline snapshot.go]: Please remove obsolete code.

**Comment by siddontang** [comment]
why not using kvproto directly?

**Comment by morefreeze** [comment]
@siddontang Next PR which I have done.

File: snapshot.go (hunk 3 above)
Comment (morefreeze) [inline snapshot.go]: Sure.

File: snapshot.go (hunk 2 above)
Comment (morefreeze) [inline snapshot.go]: Yep.

File: scan.go (hunk 1 above)
Comment (morefreeze) [inline scan.go]: How about `return GetErr()` with message from server?

Comment (disksing) [inline scan.go]: yeah, that would be reasonable.

File: error.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ticlient
+
+import "github.com/juju/errors"
+
+var (
+	// errAlreadyRollbacked return keys have been rollbacked error.
+	errAlreadyRollbacked = errors.New("already rollbacked")
```
Comment (siddontang) [inline error.go]: rolledback

File: snapshot.go (hunk 3 above)
Comment (morefreeze) [inline snapshot.go]: I have fixed in next PR.

**Comment by morefreeze** [comment]
@siddontang @disksing PTAL

**Comment by disksing** [comment]
LGTM

File: scan.go (hunk 1 above)
Comment (morefreeze) [inline scan.go]: Had fixed in next PR.

**Comment by siddontang** [comment]
Please change left rollbacked to rolled back.

File: kvrpc/proto/kvrpcpb.proto [hunk 5]
```diff
@@ -58,13 +65,22 @@ message CmdPrewriteRequest {
...
+    // primary_lock_key
+    optional bytes primary_lock = 4;
+    optional uint64 start_version = 5;
 }
 
 message CmdPrewriteResponse {
     optional bool ok = 1;
-    optional bytes primary_lock = 2;
-    optional uint64 lock_version = 3;
+    message Item {
```
Comment (siddontang) [inline kvrpc/proto/kvrpcpb.proto]: no need to extract a outer Item for both Get and Prewrite Response?

File: txn.go [hunk 6]
```diff
@@ -118,7 +133,7 @@ func (txn *tikvTxn) doCommit() error {
 		return errors.Trace(err)
 	}
 	if resp.GetType() != pb.MessageType_CmdPrewrite {
-		return errors.Errorf("receive response[%s] dismatch request[%s]",
+		return errors.Errorf("receive response[%s] mismatch request[%s]",
```
Comment (siddontang) [inline txn.go]: mismatches

File: txn.go [hunk 7]
```diff
@@ -143,12 +179,13 @@ func (txn *tikvTxn) doCommit() error {
...
 
 	// Receive commit response.
+	log.Debug("Start to commit", startTS, commitTS)
 	commitResp, err := txn.store.client.Send(commitReq)
 	if err != nil {
 		return errors.Trace(err)
 	}
 	if commitResp.GetType() != pb.MessageType_CmdCommit {
-		return errors.Errorf("receive response[%s] dismatch request[%s]",
+		return errors.Errorf("receive response[%s] mismatch request[%s]",
```
Comment (siddontang) [inline txn.go]: mismatches

File: kvrpc/proto/kvrpcpb.proto (hunk 5 above)
Comment (morefreeze) [inline kvrpc/proto/kvrpcpb.proto]: @siddontang Work is more but benefit is less. So extract them if they appear 3 times or more.

File: lock.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+	}
+}
+
+// TTL time to live
+const TTL uint64 = 10 // seconds
+// cleanup cleanup lock.pk
+// if lock.getReq != nil, call rollback&get or commit&get
+func (l *txnLock) cleanup() ([]byte, error) {
+	if !l.store.oracle.IsExpired(l.pl.version, TTL) {
+		time.Sleep(time.Since(time.Unix(int64(l.pl.version), 0)))
```
Comment (siddontang) [inline lock.go]: l.pl.version is > time.Now or not?

**Comment by morefreeze** [comment]
@siddontang PTAL

**Comment by siddontang** [comment]
LGTM

**Pull request merged**